File: src/algorithm/leetcode/problem32.java
package algorithm.leetcode;

import java.util.Stack;

public class problem32 {
    public static int longestValidParentheses(String s) {
        Stack<Integer> stack = new Stack<>();
        int max = 0;
        int start = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            } else {
                if (!stack.empty()) {
                    int p = stack.pop();
                    max = stack.empty() ? Math.max(max, i - start + 1) : Math.max(max, i - stack.peek());
                } else {
                    start = i + 1;
                }
            }
        }
        return max;
    }

    public static void main(String[] args) {
        System.out.println(longestValidParentheses("(()(())"));
    }
}

File: src/algorithm_cpp/problem32.cpp
#include<string>
#include<iostream>

using namespace std;

class Solution {
public:
    int longestValidParentheses(string s) {
	int dp[s.length()];
	int max = 0;
	for (int i = 0; i < s.length(); i++)
	    dp[i] = 0;
	// i是dp index
	for (int i = 1; i < s.length(); i++) {
	    if (s.at(i) == ')' && s.at(i - 1) == '(') {
		if (i >= 2)
		    dp[i] = dp[i - 2] + 2;
		else
		    dp[i] = 2;
	    } else if (s.at(i) == ')' && s.at(i - 1) == ')') {
		if (i >= dp[i - 1] + 1 && s.at(i - dp[i - 1] - 1) == '(') {
		    if (i >= dp[i - 1] + 2)
			dp[i] = dp[i - 1] + dp[i - dp[i - 1] - 2] + 2;
		    else 
			dp[i] = dp[i - 1] + 2;
		}
	    }
	    if (dp[i] > max)
		max = dp[i];
	}
	return max;
    }
};

int main() {
    Solution s;
    cout << s.longestValidParentheses("(()");
}
